Include what requires upgrade in upgrade message

kaapstorm opened the pull request:
`area.find('*').hide();` was hiding the form element that required the upgrade. Including the form element text allows the user to understand what it is that the upgrade will provide.

This is for [FB 162462](http://manage.dimagi.com/default.asp?162462).

![upgrade_message](https://cloud.githubusercontent.com/assets/708421/6778792/8d1e9f12-d15e-11e4-8140-7e75df14d6b9.png)

kaapstorm committed:
Include what requires upgrade in upgrade message

`area.find('*').hide();` was hiding the form element that required the
upgrade. Including the form element text allows the user to understand what it
is that the upgrade will provide.

czue commented:
thanks. I think @amsagoff might have a sense of whether we want to show this by default (it seems pretty confusing to me). either way, I think we definitely want to hide all references to module filtering at least until 2.20 is out so recommend putting this message behind the toggle. @benrudolph might have some pointers for you.

benrudolph commented:
Yea I was just looking at this PR to see what the best way of doing this was. Turns out I didn't use a toggle for this particular feature (I did for form filtering). Makes me feel like the `data-since-version` thing could benefit from some additional functionality or something because it is a useful expression.

If you want to hide it behind a toggle, it should be fairly trivial. You'd just need to add a toggle and then a few if statements around relevant code. Here's the PR: https://github.com/dimagi/commcare-hq/pull/5828/files. In case you haven't used toggles in the template yet, there's a nice templatetag `request|toggle_enabled:MY_TOGGLE_NAME`

kaapstorm committed:
Merge branch 'master' into upgrade_message

kaapstorm committed:
Preserve original behaviour

Move commcare-feature attributes from input field div to parent div.

kaapstorm committed:
Hide vaporware features

Pass latest CommCare version to template, and check commcare-feature-tagged
items.

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -26,10 +33,19 @@
     };
+    COMMCAREHQ.app_manager.versionGT = function (commcareVersion1, commcareVersion2) {
+        if (commcareVersion1 === commcareVersion2) {
+            return false
```
Missing semicolon.

kaapstorm committed:
Missing semicolon

kaapstorm commented:
@snopoke , @czue, @benrudolph Based on Simon's feedback on this, I took a different approach from my first attempt.

I've added two commits (and a third which was literally just a semicolon). The [first commit][1] brings this bug fix in line with functionality elsewhere -- the problem was that the label was being hidden instead of just the div that contained the input field.

The [second commit][2] probably needs some eyeballs. If the feature is vaporware, it hides the whole form item (i.e. the outer div). It figures out what is vaporware by taking the last CommCare build version. The view passes it in the context. 

Currently only the module view uses it. If it's added to the context for the app view, form views and releases view it should just work. 

If latest_commcare_version is not set in the context, it falls back to "0" [when app_manager is initialized][3]. [Feature version checking][4] now includes checkCommcareVaporWare, which [returns false][5] if the version is "0", and true if the feature version is greater than the latest version.


  [1]: https://github.com/dimagi/commcare-hq/commit/dcc0d8939c86243974aafec34e3b9a7bcf3d2b2d
  [2]: https://github.com/dimagi/commcare-hq/commit/ec0fd9a27e30a3427eb9228b3428eab01d708599
  [3]: https://github.com/dimagi/commcare-hq/pull/5989/files#diff-b9f3e123b20e0eb48bc2a538c98b2fc6R56
  [4]: https://github.com/dimagi/commcare-hq/pull/5989/files#diff-0f3de4bc2a0b5746ed5f36e006fdf66aR271
  [5]: https://github.com/dimagi/commcare-hq/pull/5989/files#diff-0f3de4bc2a0b5746ed5f36e006fdf66aR14

czue commented:
are you using the word "vaporware" to mean features that aren't released yet? i find that terminology a little confusing if so since i normally associate that with things that aren't real at all.

concept sounds great though and happy to merge if @snopoke or @benrudolph is happy. I think we should try to get this in today if possible since the arrow is a bit embarassing

benrudolph commented:
agree with cory about the vaporware thing. i think this should definitely get in for this deploy. maybe an update after this has been merged?

benrudolph commented on corehq/apps/app_manager/static/app_manager/js/app_manager.js:
```diff
@@ -9,6 +9,13 @@
         return COMMCAREHQ.app_manager.versionGE(COMMCAREHQ.app_manager.commcareVersion(), version);
     };
+    COMMCAREHQ.app_manager.checkCommcareVaporWare = function (version) {
+        if (COMMCAREHQ.app_manager.latestCommcareVersion() === "0") {
```
This could now change to `!COMMCAREHQ.app_manager.latestCommcareVersion()` if we use `null` as the way of denoting that the version is absent

kaapstorm committed:
Use null instead of "0", and "vapor" is misleading

kaapstorm commented:
Thanks for all the feedback @czue, @benrudolph. I've changed "0" to null: Much nicer. And changed "VaporWare" (you're right Cory; vaporware never arrives) to "AreWeThereYet" -- and checking for the inverse of checkVaporWare (i.e checking our_version >= feature_version) also allows us to drop that versionGT method.

benrudolph commented:
Thanks for the changes @kaapstorm. I'm going to merge this for now so the fix will be out for deploy. I still think we may want to rethink `parent` since if this was the current functionality it would have hid the whole form which would have been a much worse UI bug

amsagoff commented:
Just making sure this is still under a flag and *not* going to be exposed
when 2.20 is released. (We are blocking product release on a productized
implementation that leverages user case data).

On Tue, Mar 24, 2015 at 2:39 PM, Ben <notifications@github.com> wrote:

> Merged #5989 <https://github.com/dimagi/commcare-hq/pull/5989>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5989#event-263864386>.
>

kaapstorm commented:
@amsagoff I'm pretty sure this *will* be exposed when 2.20 is released. @benrudolph, is this correct? Or is module filter behind a flag?

benrudolph commented:
Yes, this feature is not behind a feature flag. I can put it in my backlog
to put it behind a toggle

On Mon, Mar 30, 2015 at 3:32 PM, Norman Hooper <notifications@github.com>
wrote:

> @amsagoff <https://github.com/amsagoff> I'm pretty sure this *will* be
> exposed when 2.20 is released. @benrudolph <https://github.com/benrudolph>,
> is this correct? Or is module filter behind a flag?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5989#issuecomment-87801038>.
>

amsagoff commented:
Cool, anytime in the next 2 weeks should be good (before 2.20 goes live)

On Mon, Mar 30, 2015 at 4:09 PM, Ben <notifications@github.com> wrote:

> Yes, this feature is not behind a feature flag. I can put it in my backlog
> to put it behind a toggle
>
> On Mon, Mar 30, 2015 at 3:32 PM, Norman Hooper <notifications@github.com>
> wrote:
>
> > @amsagoff <https://github.com/amsagoff> I'm pretty sure this *will* be
> > exposed when 2.20 is released. @benrudolph <
> https://github.com/benrudolph>,
> > is this correct? Or is module filter behind a flag?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/dimagi/commcare-hq/pull/5989#issuecomment-87801038>.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/5989#issuecomment-87813451>.
>
